### ISSUE TITLE
Publisher job not running

### ISSUE THREAD

**Issue by thomastaylor312** [intelsdi-x/snap#655]
Is there any reason why a publisher job would just stop running without any error? All the data is showing up when I run `snapctl task watch`, but nothing is publishing to the influx data source. I am testing out a plugin and it was working fine until I added new metric types. But instead of failing, it is like the publish job doesn't even exist:

**Normal (before I made the change)**

```
time="2016-01-07T12:40:36-08:00" level=debug msg="collector run completed" _module=scheduler-job block=run job-type=collector metric-count=10 
time="2016-01-07T12:40:36-08:00" level=debug msg="starting publisher job" _module=scheduler-job block=run content-type=snap.gob job-type=publisher plugin-config=map[port:{8086} user:{foo} database:{t
esting} host:{example.com} password:{bar}] plugin-name=influx plugin-version=-1
time="2016-01-07T12:40:36-08:00" level=debug msg="plugin selected" _module=control-routing block=select hitcount=0 index="publisher:influx:v8:id7" pool size=1 strategy=least-recently-used 
time="2016-01-07T12:40:51-08:00" level=debug msg="task spin loop" _module=scheduler-task 
```

**Broken**

```
time="2016-01-07T12:44:53-08:00" level=debug msg="collector run completed" _module=scheduler-job block=run job-type=collector metric-count=18 
time="2016-01-07T12:44:53-08:00" level=debug msg="task spin loop" _module=scheduler-task
```

**Comment by geauxvirtual** [comment]
My initial thought is the publisher plugin is panic'ing on an error (which
is the default behavior for the InfluxDB plugin that has not been fixed
yet). Can you look at the logs for the plugin and see if anything is in
those?

On Thu, Jan 7, 2016 at 12:56 PM, Taylor Thomas <notifications@github.com>
wrote:

> Is there any reason why a publisher job would just stop running without
> any error? All the data is showing up when I run snapctl task watch, but
> nothing is publishing to the influx data source. I am testing out a plugin
> and it was working fine until I added new metric types. But instead of
> failing, it is like the publish job doesn't even exist:
>
> *Normal (before I made the change)*
>
> time="2016-01-07T12:40:36-08:00" level=debug msg="collector run completed" _module=scheduler-job block=run job-type=collector metric-count=10
> time="2016-01-07T12:40:36-08:00" level=debug msg="starting publisher job" _module=scheduler-job block=run content-type=snap.gob job-type=publisher plugin-config=map[port:{8086} user:{foo} database:{t
> esting} host:{example.com} password:{bar}] plugin-name=influx plugin-version=-1
> time="2016-01-07T12:40:36-08:00" level=debug msg="plugin selected" _module=control-routing block=select hitcount=0 index="publisher:influx:v8:id7" pool size=1 strategy=least-recently-used
> time="2016-01-07T12:40:51-08:00" level=debug msg="task spin loop" _module=scheduler-task
>
> *Broken*
>
> time="2016-01-07T12:44:53-08:00" level=debug msg="collector run completed" _module=scheduler-job block=run job-type=collector metric-count=18
> time="2016-01-07T12:44:53-08:00" level=debug msg="task spin loop" _module=scheduler-task
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/intelsdi-x/snap/issues/655>.
>



-- 
Justin Guidroz

**Comment by thomastaylor312** [comment]
This is the only thing besides heartbeats that I found:

```
2016/01/07 12:42:44 2016/01/07 12:42:44 method Decode reply type not a pointer: interface {}
2016/01/07 12:42:44 2016/01/07 12:42:44 method Decrypt has wrong number of ins: 2
2016/01/07 12:42:44 2016/01/07 12:42:44 method DecryptKey has wrong number of ins: 2
2016/01/07 12:42:44 2016/01/07 12:42:44 method Encode has wrong number of ins: 2
2016/01/07 12:42:44 2016/01/07 12:42:44 method Encrypt has wrong number of ins: 2
2016/01/07 12:42:44 2016/01/07 12:42:44 method EncryptKey has wrong number of ins: 1
2016/01/07 12:42:44 2016/01/07 12:42:44 method KillChan has wrong number of ins: 1
2016/01/07 12:42:44 2016/01/07 12:42:44 method ListenAddress has wrong number of ins: 1
2016/01/07 12:42:44 2016/01/07 12:42:44 method ListenPort has wrong number of ins: 1
2016/01/07 12:42:44 2016/01/07 12:42:44 method Logger has wrong number of ins: 1
2016/01/07 12:42:44 2016/01/07 12:42:44 method ResetHeartbeat has wrong number of ins: 1
2016/01/07 12:42:44 2016/01/07 12:42:44 method SetEncrypter has wrong number of ins: 2
2016/01/07 12:42:44 2016/01/07 12:42:44 method SetListenAddress has wrong number of ins: 2
2016/01/07 12:42:44 2016/01/07 12:42:44 method Token has wrong number of ins: 1
```

**Comment by geauxvirtual** [comment]
If you export the task via `snapctl task export`, does it match the task manifest that was loaded?

**Comment by thomastaylor312** [comment]
Yep. It is pretty much the same. The only difference was the `deadline` key, which when I made that value higher, it didn't affect anything.

**Comment by geauxvirtual** [comment]
When you loaded the new version of the plugin with your changes, was this into a fresh snapd instance, or was this into a running snapd instance with while the task was running against the old plugin?

**Comment by thomastaylor312** [comment]
I tried both. I tried with a fresh snapd instance and with a currently running snapd instance. In the second case, I unloaded the old plugin, removed the task and loaded the new plugin with a new task

**Comment by geauxvirtual** [comment]
Can you load up the plugins and task into a fresh instance of snapd and drop the entire log from the start of snapd to the first run of the task into a gist?

**Comment by thomastaylor312** [comment]
Here it is, scrubbed of some sensitive data: https://gist.github.com/thomastaylor312/c00d47e775ca759b2b6c

**Comment by geauxvirtual** [comment]
I see when this was working you were collecting 10 metrics, and now collecting 18 metrics. Are the metrics being defined in the task dynamically with a `*` or individually listed out?

**Comment by geauxvirtual** [comment]
Disregard what I just asked, though it is sort of related.

The problem revolves around what is mentioned earlier in that the exported task has a new deadline field. By default, the deadline for a job is 5 seconds. The publish job takes the parents deadline which is the time the job was created + 5 seconds. You can see here in this part of the code https://github.com/intelsdi-x/snap/blob/master/scheduler/worker.go#L49-L56, we do not generate an error log if the deadline passes, and we just kill the job. If you look at the gist you posted, 
```
time="2016-01-07T14:23:53-08:00" level=debug msg="plugin selected" _module=control-routing block=select hitcount=0 index="collector:orcs:v1:id1" pool size=1 strategy=least-recently-used 
time="2016-01-07T14:24:02-08:00" level=debug msg="collector run completed" _module=scheduler-job block=run job-type=collector metric-count=18
```
it took 9 seconds for the collection of 18 metrics to occur. Try adding to your task manifest a deadline of 15 seconds or greater and see if the job succeeds at that point.

**Comment by thomastaylor312** [comment]
They are being individually listed out.

**Comment by ConnorDoyle** [comment]
Created an issue (#656) to track the missing log messages.

**Comment by thomastaylor312** [comment]
I did try that. I even moved it up to 60s and it still didn't work.

**Comment by geauxvirtual** [comment]
Looking through the code, while it looks like we have a way to change deadline duration, we haven't exposed getting the setting from the task manifest or from the command line. That's my initial hunch, and will continue to look into this tonight.

**Comment by thomastaylor312** [comment]
Awesome. Thanks for looking into this @geauxvirtual and @ConnorDoyle Let me know if you need any more info from me.

**Comment by geauxvirtual** [comment]
Found the issue why adding deadline to the task manifest is not working. We have deadline support on the server side of the REST api, but the client and the command line tool currently do not support the deadline option from the task manifest as seen here https://github.com/intelsdi-x/snap/blob/master/cmd/snapctl/task.go#L86-L91 and here https://github.com/intelsdi-x/snap/blob/master/mgmt/rest/client/task.go#L50-L58.

I'll start working on a fix tonight, and try to get that in to master in the next day or so.

**Comment by geauxvirtual** [comment]
PR #657 has been merged into master that should allow setting deadline in the task manifest, and it actually be followed. Pull down the latest master and see if that resolves your current issue.

**Comment by thomastaylor312** [comment]
Yep this is working now.

**Comment by thomastaylor312** [comment]
Closed by #658
